docs/notes/parameters.md

```diff
@@ -0,0 +1,89 @@
+# Design note: dealing with Module paramaters
+
+## Introduction
+
+In linen we can specify `Module` paramaters either as attributes or method arguments (usually `__call__`).
+Usually the distinction is obvious: Hyperparamaters like the number of filters or the kernel initializer
+are hyper paramaters and call time arguments are the data and some "mode switches" like whether the model
+should be deterministic or use stochastic regularization.
+
+Some cases are however ambigious. Take for example the `Dropout` module.
+We have a number of clear hyper paramaters:
+1. the dropout rate
+2. the axes for which a dropout mask is generated
+
+and some clear call time arguments:
+1. the input that should be masked using dropout
+2. the (optional) rng used to sample the random mask
+
+There is however a final argument that is ambigious. Which is the `deterministic` argument.
+If `deterministic` is true no dropout mask is sampled. This is typically used during model evaluation.
+However, if we pass `eval=True` or `train=False` to a top-level Module. The `deterministic` argument needs
+to be applied everywhere and the boolean argument needs to be passed down to all the layers that might use `Dropout`.
+If `deterministic` is an attribute we might do the following instead:
+
+```python
+from functools import partial
+from flax import linen as nn
+
+class ResidualModel(nn.Module):
+  drop_rate: float
+
+  @nn.compact
+  def __call__(self, x, *, train):
+    dropout = partial(nn.Dropout, rate=self.drop_rate, deterministic=not train)
+    for i in range(10):
+      x += ResidualBlock(dropout=dropout, ...)(x)
+```
+
+It makes sense to pass `determinstic` to the constructor here because this way we can pass the dropout template to the sub-modules.
+Now the sub-module no longer needs to take care of train vs eval mode and can simply use the `dropout` argument.
+Note that because the dropout layer can only be constructed in the sub-module we can only partially `deterministic` to the constructor but not to `__call__`.
+
+However, if deterministic is just an attribute we run into trouble when using the setup pattern:
+
+```python
+class SomeModule(nn.Module):
+  drop_rate: float
+
+  def setup(self):
+    self.dropout = nn.Dropout(rate=self.drop_rate)
+
+  @nn.compact
+  def __call__(self, x, *, train):
+    # ...
+    x = self.dropout(x, deterministic=not train)
+    # ...
+```
+
+Here it makes sense to pass `deterministic` during `__call__` because it depends on the `train` argument.
```

Comment:
```suggestion
But as defined above `deterministic` would be an attribute, so this doesn't work. Here it makes sense to pass `deterministic` during `__call__` because it depends on the `train` argument.
```